File: trees_and_graphs/find_largest_value_in_each_tree_row.js
/**
 * @param {TreeNode} root
 * @returns {number[]}
 */
var largestValues = function(root) {
    if (!root)
    {
        return [];
    }

    let ans = [];
    let queue = [root];

    while (queue.length)
    {
        const currentLength = queue.length;
        let currMax = Number.MIN_SAFE_INTEGER;
        // let currMax = -Infinity;
        let nextQueue = [];

        for (let i = 0; i < currentLength; i++)
        {
            const node = queue.shift();
            // const node = queue[i];
            currMax = Math.max(currMax, node.val);

            if (node.left)
            {
                nextQueue.push(node.left);
            }

            if (node.right)
            {
                nextQueue.push(node.right);
            }
        }

        ans.push(currMax);
        queue = nextQueue;
    }

    return ans;
};

File: trees_and_graphs/find_largest_value_in_each_tree_row.cpp
#include "treenode.hpp"

#include <gtest/gtest.h>

#include <algorithm>
#include <limits>
#include <vector>
#include <queue>

//! @brief Find largest value in each row of the tree
//! @param[in] root Pointer to root of binary tree
//! @return Vector of the largest value in each row of the tree
static std::vector<int> largestValues(TreeNode* root)
{
    if (root == nullptr)
    {
        return {};
    }

    std::vector<int> ans {};
    std::queue<TreeNode*> queue {};
    queue.push(root);

    while (not queue.empty())
    {
        const auto currentLength = static_cast<int>(queue.size());
        int        currMax {std::numeric_limits<int>::min()};

        for (int i = 0; i < currentLength; ++i)
        {
            const auto node = queue.front();
            currMax         = std::max(currMax, node->val);
            queue.pop();

            if (node->left != nullptr)
            {
                queue.push(node->left);
            }

            if (node->right != nullptr)
            {
                queue.push(node->right);
            }
        }

        ans.push_back(currMax);
    }

    return ans;

} // static std::vector<int> largestValues( ...

TEST(LargestValuesTest, SampleTest)
{
    TreeNode one {1};
    TreeNode two {2};
    TreeNode three_up {3};
    
    const TreeNode three_down {3};
    const TreeNode five {5};
    const TreeNode nine {9};

    one.left  = &three_up;
    one.right = &two;

    three_up.left  = &five;
    three_up.right = &three_down;

    two.right = &nine;

    const std::vector<int> expected_output {1, 3, 9};
    EXPECT_EQ(expected_output, largestValues(&one));
}